### DISCUSSION TITLE
Restore data after update.

### DISCUSSION THREAD

**Discussion by Sieg-hart** [teslamate-org/teslamate#4422] [Q&A] [answered]
Hello,

Sorry if this question has already been asked.

I recently upgraded from version _1.30.0_ to version _1.32.0_.

I made a backup of the database using this documentation, via the command `docker-compose exec -T database pg_dump -U teslamate teslamate > ./teslamate.bck`

I have a 60mb _teslamate.bck_ file that seems to contain all the data.

I deleted the stack containing teslamate and recreated a new stack from scratch.

I now have a teslamate instance in _1.32.0_ with a postgreSQL database version 17.

The problem is that when I try to restore the data... nothing happens.
I run the command `docker-compose exec -T database psql -U teslamate -d teslamate < teslamate.bck`.
I get the prompt without any message.
```
root@raspberrypi:/home/sieg# docker-compose exec -T database psql -U teslamate -d teslamate < teslamate.bck
root@raspberrypi:/home/sieg#
```

When I look in Grafana, I have no data.

And when I look at the database logs, I have error messages related to SQL syntax errors but in queries that are not in my .bck.

For example :
2024-12-02 22:49:12.203 UTC [90] ERROR: syntax error at or near ")" at character 295
2024-12-02 22:49:12.203 UTC [90] STATEMENT: SELECT percentile_disc(0.5) WITHIN GROUP (ORDER BY since_last_update) FROM (
 SELECT extract(EPOCH FROM start_date - lag(start_date) OVER (ORDER BY start_date)) AS since_last_update
 FROM updates
 WHERE start_date BETWEEN '2014-12-02T22:49:13.069Z' AND '2024-12-02T22:49:13.069Z' AND car_id =
 ) d;
2024-12-02 22:49:12.241 UTC [91] ERROR: syntax error at or near "and" at character 145

There is no such select in my file, just `create`, and `copy`.

I tried to execute the SQL queries directly in the database.
The "creates" worked (with a few ; less).
The "copies" didn't work.
I tried to transform the copies into insert into, some data is added to the database, but does not appear in Grafana.

I don't know what to do anymore. :(

**Comment by eliluong** [comment]
What was the old Postgres version? What you may need to do is stop all containers, delete the "new" database volume that is not working. Then start only the database container. Then try the restore command here: https://docs.teslamate.org/docs/maintenance/backup_restore

If that does not work, you may have to get into the Docker database container (something like https://stackoverflow.com/questions/30172605/how-do-i-get-into-a-docker-containers-shell), then use chatgpt or something similar to help you format the commands correctly. You will probably have copy the backup file into the container. Then you can see whatever error messages show up during import.

I was on Postgres 13 for a long time before upgrading to Postgres 15, and it was not straightforward.

**Reply by Sieg-hart** [reply]
Hi,

it was the version 16 of PostgreSQL.

I deleted everything (stack and database volume) and reinstalled everything from scratch.

I mapped the database so that it is accessible from my pc.
I modified the .bck file to run it from my pc (I transformed the `copy` into `insert into`).
I did not get any error and I see the data in the database.

I restarted all containers but I do not see anything in grafana.


After that, i reconfigured Teslamate to retrieve data from my car.
I can see the new data clearly (but still not the old data), and the weird thing is that I don't see the new data anywhere in the database...

And I have sone new error on the database log

```
2024-12-03 15:43:57.752 UTC [1] LOG:  database system is ready to accept connections
2024-12-03 15:45:40.247 UTC [47] ERROR:  invalid input syntax for type json at character 41
2024-12-03 15:45:40.247 UTC [47] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.247 UTC [47] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.247 UTC [47] STATEMENT:  SELECT  
	  CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxCapacity')::float END as "Usable (new)", 
	  (''::json ->> 'CurrentCapacity')::float as "Usable (now)",
	  (''::json ->> 'CurrentCapacity')::float - CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxCapacity')::float END as "Difference"
	  
	 
2024-12-03 15:45:40.266 UTC [46] ERROR:  invalid input syntax for type json at character 52
2024-12-03 15:45:40.266 UTC [46] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.266 UTC [46] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.266 UTC [46] STATEMENT:  SELECT  
	  LEAST(100, (100 - GREATEST(0, 100.0 - ((''::json ->> 'CurrentCapacity')::float * 100.0 / CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxCapacity')::float END)))) as "Battery Health (%)"
	  
	
2024-12-03 15:45:40.318 UTC [52] ERROR:  invalid input syntax for type json at character 71
2024-12-03 15:45:40.318 UTC [52] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.318 UTC [52] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.318 UTC [52] STATEMENT:  SELECT
		floor(sum(charge_energy_added) / CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxCapacity')::float END) AS "# of Charging cycles"
	FROM charging_processes WHERE car_id = 1 AND charge_energy_added > 0.01
2024-12-03 15:45:40.322 UTC [51] ERROR:  invalid input syntax for type json at character 9
2024-12-03 15:45:40.322 UTC [51] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.322 UTC [51] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.322 UTC [51] STATEMENT:  SELECT (''::json ->> 'RatedEfficiency')::float * 10 / convert_km(1, 'km') AS efficiency_km
2024-12-03 15:45:40.329 UTC [52] ERROR:  invalid input syntax for type json at character 41
2024-12-03 15:45:40.329 UTC [52] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.329 UTC [52] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.329 UTC [52] STATEMENT:  SELECT  
	  CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxRange')::float END as "maxrange_km",
	  (''::json ->> 'CurrentRange')::float  as "currentrange_km",
	  CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxRange')::float END - (''::json ->> 'CurrentRange')::float as "range_lost_km"
2024-12-03 15:45:40.329 UTC [51] ERROR:  invalid input syntax for type json at character 48
2024-12-03 15:45:40.329 UTC [51] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.329 UTC [51] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.329 UTC [51] STATEMENT:  SELECT * FROM ((SELECT usable_battery_level * (''::json ->> 'CurrentCapacity')::float / 100 as kWh, date, (''::json ->> 'CurrentCapacity')::float as Total
	FROM positions
	WHERE car_id = 1  AND usable_battery_level IS NOT NULL
	ORDER BY date DESC

)
	UNION
	(SELECT battery_level * (''::json ->> 'CurrentCapacity')::float / 100 as kWh, date, (''::json ->> 'CurrentCapacity')::float as Total
	FROM charges c
	JOIN charging_processes p ON p.id = c.charging_process_id
	WHERE p.car_id = 1  AND usable_battery_level IS NOT NULL
	ORDER BY date DESC
	LIMIT 1)) AS last_usable_battery_level LIMIT 1
2024-12-03 15:45:40.341 UTC [52] ERROR:  invalid input syntax for type json at character 100
2024-12-03 15:45:40.341 UTC [52] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.341 UTC [52] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.341 UTC [52] STATEMENT:  SELECT convert_km(AVG(p.odometer)::numeric,'km') AS "Odometer", 
		AVG(c.rated_battery_range_km * (''::json ->> 'RatedEfficiency')::float / c.usable_battery_level) AS "kWh",
		--MAX(cp.id) AS id,
		to_char(cp.end_date, 'YYYY-MM-dd') AS "Date"
		FROM charging_processes cp
			JOIN (SELECT charging_process_id, MAX(date) as date	FROM charges WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges	ON cp.id = last_charges.charging_process_id
			INNER JOIN charges c
			ON c.charging_process_id = cp.id AND c.date = last_charges.date
			INNER JOIN positions p ON p.id = cp.position_id
		WHERE cp.car_id = 1
			AND cp.end_date IS NOT NULL
			AND cp.charge_energy_added >= (''::json ->> 'RatedEfficiency')::float


		GROUP BY 3
2024-12-03 15:45:40.341 UTC [54] ERROR:  invalid input syntax for type json at character 155
2024-12-03 15:45:40.341 UTC [54] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.341 UTC [54] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.341 UTC [54] STATEMENT:  SELECT 
	  ROUND(MIN(convert_km(p.odometer::numeric,'km')),0) AS "Odometer",
		ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP(ORDER BY c.rated_battery_range_km * (''::json ->> 'RatedEfficiency')::float / c.usable_battery_level)::numeric,1) AS "kWh",
		to_char(cp.end_date, 'YYYYMM') || CASE WHEN to_char(cp.end_date, 'DD')::int <= 15 THEN '1' ELSE '2' END  AS Title
		FROM charging_processes cp
			JOIN (SELECT charging_process_id, MAX(date) as date	FROM charges WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges	ON cp.id = last_charges.charging_process_id
			INNER JOIN charges c
			ON c.charging_process_id = cp.id AND c.date = last_charges.date
			INNER JOIN positions p ON p.id = cp.position_id
		WHERE cp.car_id = 1
			AND cp.end_date IS NOT NULL
			AND cp.charge_energy_added >= (''::json ->> 'RatedEfficiency')::float

		GROUP BY 3
2024-12-03 15:45:40.354 UTC [53] ERROR:  invalid input syntax for type json at character 30
2024-12-03 15:45:40.354 UTC [53] DETAIL:  The input string ended unexpectedly.
2024-12-03 15:45:40.354 UTC [53] CONTEXT:  JSON data, line 1: 
2024-12-03 15:45:40.354 UTC [53] STATEMENT:  SELECT GREATEST(0, 100.0 - ((''::json ->> 'CurrentCapacity')::float * 100.0 / CASE WHEN 0 > 0 THEN 0 ELSE (''::json ->> 'MaxCapacity')::float END))
```

**Reply by brianmay** [reply]
@Sieg-hart Are you sure you are were on postgresql 16? This really does look like errors I might have expected from dump file being in an incompatible format.

If you are able to confirm that you did have an earlier postgresql, and you no longer have the original data, I would suggest restoring to the exact same version as the first step.

**Reply by Sieg-hart** [reply]
The first time I tried to update, I only deleted the stack, not the volume.
I had errors in the database logs telling me that I had version 16 and that I needed version 17.
Following this, I deleted the volume and modified the docker-compose.yml file to indicate that version 17 of the DB was required.
```
database:
    image: postgres:17
```

**Reply by brianmay** [reply]
Oh, oops, sorry, was falling behind the times. Thought 16 was the latest for some reason.

I think I might be trying to restore to 16 first, and then upgrading.

Not sure about the errors that you needed version 17. I am still running 16 on all of my databases, and teslamate seems to be fine with it. Guess I should fix that sometime.

**Reply by Sieg-hart** [reply]
I recreated a stack but with postgreSQL version 16.

I re-imported the data but grafana still doesn't see anything.

Is it really in the "public" base of the Teslamate database that the data must be imported?

I don't understand anything about it.
the data is there but grafana does not see it.

And when I reconfigure the API to retrieve the data from my car, I see the data in grafana, but not in the database...

And in the database's log, i always have the same SQL query errors.

**Reply by brianmay** [reply]
Are you sure grafana is looking at the correct database?

**Reply by eliluong** [reply]
I think you have no data showing in Grafana because the log you posted indicate that (probably) Grafana is querying the database, but there is an issue with the underlying data. All those `SELECT` statement I think are queries that Grafana does on the database. If you check the dev console or Grafana container when you load a Grafana page, are there errors?

I'm not sure if you can create a new backup, but that would be easiest.

I would delete all the containers/volumes. Start fresh, only start up the database container, and then try restoring the backup inside the container. It might show more informative information if it runs into an error.

**Reply by Sieg-hart** [reply]
> Are you sure grafana is looking at the correct database?

This is the question I'm asking myself, but I only have a postgreSQL database on my raspberry.
I connect to the teslamate-database-1 database, with the login and password, and to the port indicated in docker-compose.yml.
I don't see any other database. :/

```
root@raspberrypi:/home/sieg# docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED        STATUS                  PORTS                                                                                                               NAMES
f19037177bf2   postgres:16                     "docker-entrypoint.s…"   15 hours ago   Up 15 hours             0.0.0.0:5432->5432/tcp, :::5432->5432/tcp                                                                           teslamate-database-1
c63ce45919b9   teslamate/grafana:latest        "/run.sh"                15 hours ago   Up 14 hours             0.0.0.0:3000->3000/tcp, :::3000->3000/tcp                                                                           teslamate-grafana-1
b27eb5892ee9   eclipse-mosquitto:2             "/docker-entrypoint.…"   15 hours ago   Up 14 hours             1883/tcp                                                                                                            teslamate-mosquitto-1
5accd86da191   teslamate/teslamate:latest      "tini -- /bin/sh /en…"   15 hours ago   Up 14 hours             0.0.0.0:4000->4000/tcp, :::4000->4000/tcp                                                                           teslamate-teslamate-1
a39925d02d5f   portainer/portainer-ce:2.21.4   "/portainer"             2 days ago     Up 21 hours             0.0.0.0:8000->8000/tcp, :::8000->8000/tcp, 0.0.0.0:9443->9443/tcp, :::9443->9443/tcp, 9000/tcp                      portainer
cb70dec2d0d8   pihole/pihole:latest            "/s6-init"               2 months ago   Up 21 hours (healthy)   0.0.0.0:53->53/tcp, 0.0.0.0:53->53/udp, :::53->53/tcp, :::53->53/udp, 67/udp, 0.0.0.0:800->80/tcp, :::800->80/tcp   pihole
```

**Reply by Sieg-hart** [reply]
> I think you have no data showing in Grafana because the log you posted indicate that (probably) Grafana is querying the database, but there is an issue with the underlying data. All those `SELECT` statement I think are queries that Grafana does on the database. If you check the dev console or Grafana container when you load a Grafana page, are there errors?
> 
> I'm not sure if you can create a new backup, but that would be easiest.
> 
> I would delete all the containers/volumes. Start fresh, only start up the database container, and then try restoring the backup inside the container. It might show more informative information if it runs into an error.

I have no error message in the dev console, and in the grafana logs I only have lines `msg="Request Completed"`, `msg="Completed cleanup jobs"` or `msg="Usage stats are ready to report"`

I deleted my initial stack, I can no longer create a new backup.

I wanted to look at what the new data was like to possibly reformat the old ones if the model had changed but I don't see the new data in the database.

I reinstalled the stack from scratch, without trying to integrate my old data.
I put the requested information to connect to the tesla api.
I see the data showing on the web interface on port 4000, as well as in grafana. so far, nothing unusual.

But still nothing in the database, the public database remains empty, and I still have SQL errors in the database logs.

**Reply by Sieg-hart** [reply, marked as answer]
I think I found the problem.

Unable to import the data via the command lines, I modified the dump to change the `copy` to `insert into` (using a custom script), and modified the function declarations to remove the ; after `END`.

By exemple :
```
CREATE FUNCTION public.convert_m(n double precision, unit text) RETURNS double precision
    LANGUAGE sql IMMUTABLE STRICT
    AS $_$
  SELECT
    CASE WHEN $2 = 'm' THEN $1
         WHEN $2 = 'ft' THEN $1 * 3.28084
    END
$_$;
```

And I tried to connect to the database from outside.
The tool I used (heidiSQL) did not show me all the databases (the tool seems compatible more with mySQL databases than postgreSQL.).
I only saw the "postgres" base and not "teslamate".

So I put the data in the wrong database.

So I recreated the stack from scratch, stopped all the containers except the database, (important otherwise the tables and foreign keys are created and prevent the addition of data) I connected to it with pgAdmin4. By running the dump, table by table, on the correct base, I was able to add my old data (except the contents of the token table which generated errors).

After restarting the containers and configuring the API tokens, **it works!**

Thank you for your help.

**Comment by brianmay** [comment]
My limited experience was that plain sql files don't work so well when changing psql versions.

e.g. not tested.

```
pg_dumpall --format tar > db,dump
pg_restore --create --format tar db.dump
```

Wonder if we need to update the docs.
